src/aio.rs

```diff
@@ -403,180 +394,168 @@ where
-        try_ready!(self.sink_stream.poll_complete().map_err(|err| {
-            self.send_result(Err(err));
-        }));
-        self.poll_read()
+    fn poll_flush(
+        mut self: Pin<&mut Self>,
+        cx: &mut task::Context,
+    ) -> Poll<Result<(), Self::Error>> {
+        ready!(self
+            .as_mut()
+            .project()
+            .sink_stream
+            .poll_flush(cx)
+            .map_err(|err| {
+                self.as_mut().send_result(Err(err));
+            }))?;
+        self.poll_read(cx).map(Ok)
     }
 
-    fn close(&mut self) -> Poll<(), Self::SinkError> {
+    fn poll_close(
+        mut self: Pin<&mut Self>,
+        cx: &mut task::Context,
+    ) -> Poll<Result<(), Self::Error>> {
         // No new requests will come in after the first call to `close` but we need to complete any
         // in progress requests before closing
         if !self.in_flight.is_empty() {
-            try_ready!(self.poll_complete());
+            ready!(self.as_mut().poll_flush(cx))?;
         }
-        self.sink_stream.close().map_err(|err| {
+        let this = self.as_mut().project();
+        this.sink_stream.poll_close(cx).map_err(|err| {
             self.send_result(Err(err));
         })
     }
 }
 
-impl<T> Pipeline<T>
+impl<SinkItem, I, E> Pipeline<SinkItem, I, E>
 where
-    T: Sink<SinkError = <T as Stream>::Error> + Stream + Send + 'static,
-    T::SinkItem: Send,
-    T::Item: Send,
-    T::Error: Send,
-    T::Error: ::std::fmt::Debug,
+    SinkItem: Send + 'static,
+    I: Send + 'static,
+    E: Send + 'static,
 {
-    fn new<E>(sink_stream: T, executor: E) -> Self
+    fn new<T, F>(sink_stream: T, mut executor: F) -> Self
     where
-        E: Executor<Box<dyn Future<Item = (), Error = ()> + Send>>,
+        F: Spawn,
+        T: Sink<SinkItem, Error = E> + Stream<Item = Result<I, E>> + 'static,
+        T: Send + 'static,
+        T::Item: Send,
+        T::Error: Send,
+        T::Error: ::std::fmt::Debug,
     {
         const BUFFER_SIZE: usize = 50;
         let (sender, receiver) = mpsc::channel(BUFFER_SIZE);
         let f = receiver
-            .map_err(|_| ())
-            .forward(PipelineSink {
-                sink_stream,
-                in_flight: VecDeque::new(),
-            })
+            .map(Ok)
+            .forward(PipelineSink::new::<SinkItem>(sink_stream))
             .map(|_| ());
-
-        executor.execute(Box::new(f)).unwrap();
+        executor.spawn(f).expect("Unable to spawn redis task");
         Pipeline(sender)
     }
 
     // `None` means that the stream was out of items causing that poll loop to shut down.
-    fn send(
-        &self,
-        item: T::SinkItem,
-    ) -> impl Future<Item = T::Item, Error = Option<T::Error>> + Send {
+    async fn send(&mut self, item: SinkItem) -> Result<I, Option<E>> {
         self.send_recv_multiple(item, 1)
-            .map(|mut item| item.pop().unwrap())
+            .map_ok(|mut item| item.pop().unwrap())
```

Comment:
(Not directly related to your PR, but anyway:) is this `unwrap()` safe? Maybe add some comment explaining why this is okay?